presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -858,24 +859,44 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<C
             }
 
             List<TypeSignatureProvider> argumentTypes = getCallArgumentTypes(node.getArguments(), context);
-            Signature function = resolveFunction(node, argumentTypes, metadata);
+
+            Signature function;
+            try {
+                function = metadata.resolveFunction(node.getName(), argumentTypes);
+            }
+            catch (PrestoException e) {
+                if (e.getLocation().isPresent()) {
+                    // If analysis of any of the argument types (which is done lazily to deal with lambda expressions) fails, we
+                    // want to report the original reason for the failure
+                    throw e;
+                }
+
+                // otherwise, it must have failed due to a missing function or other reason, so we report an error at the
+                // current location
+                throw new PrestoException(
```

Comment:
Maybe introduce a variable to avoid excessive wrapping (also makes the throw easier to read)
```java
Location location = node.getLocation()
        .map(location -> new Location(location.getLineNumber(), location.getColumnNumber()));
throw new PrestoException(e::getErrorCode, location, e.getMessage(), e);

Reply:
(It's actually an optional, so it ends up being slightly more verbose):

```java
Optional<Location> errorLocation = node.getLocation()
        .map(location -> new Location(location.getLineNumber(), location.getColumnNumber()));
throw new PrestoException(e::getErrorCode, errorLocation, e.getMessage(), e);
```

I'm not sure that ends up being much better. Yes, it removes wrapping, but there's an additional variable to think about (and the fact it's there makes it seem "important"). A better way would be to introduce a method for that translation:

```java
throw new PrestoException(e::getErrorCode, extractLocation(node), e.getMessage(), e);
```